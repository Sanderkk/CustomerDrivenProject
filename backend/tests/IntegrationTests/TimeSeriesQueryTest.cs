using HotChocolate;
using HotChocolate.Execution;
using HotChocolate.Types;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using src.Api.Inputs;
using src.Api.Queries;
using src.Config;
using src.Database;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace tests.IntegrationTests
{
    public class TimeSeriesQueryTest
    {

        private string dbConnectionString = "Host=sanderkk.com;Username=sintef;Password=123456;Database=fishfarm2";

        [Fact]
        public async Task TestTimeSeriesQuery()
        {
            IServiceProvider serviceProvider =
                new ServiceCollection()
                .AddSingleton<IFishFarmRepository, FishFarmRepository>()
                .AddSingleton<IDatabaseConfig>(sp =>
                new DatabaseConfig()
                {
                    DatabaseConnectionString = dbConnectionString
                }
                )
                .BuildServiceProvider();

            IQueryExecutor executor = Schema.Create(c =>
            {
                c.RegisterQueryType(new ObjectType<TimeSeriesQuery>(d => d.Name("Query")));
                c.RegisterType<GenericObject>();
            })
            .MakeExecutable();

            var input = new TimeSeriesRequestInput()
            {
                Sensors = new List<int>() { 2, 3 },
                SpecifiedTimePeriode = true,
                From = DateTime.Parse("2020-08-12T08:21:19.000Z"),
                To = DateTime.Parse("2020-08-17T12:47:21.000Z")
            };

            IReadOnlyQueryRequest request =
                QueryRequestBuilder.New()
                .SetQuery(@"query TimeSeries($input: TimeSeriesRequestInput!){
                              timeSeries(input: $input) {
                                table,
                                startDate,
                                endDate,
                                data {
                                    name,
                                    startTime
                                    interval
                                    data
                                },
                                time
                              }
                            }")
                .SetServices(serviceProvider)
                .AddVariableValue("input", input)
                .Create();

            // act
            IExecutionResult result = await executor.ExecuteAsync(request);
            //Snapshot.Match(result);
            result.MatchSnapshot();
        }

        [Fact]
        public async Task TestTimeSeriesQueryBackwards()
        {
            IServiceProvider serviceProvider =
                new ServiceCollection()
                .AddSingleton<IFishFarmRepository, FishFarmRepository>()
                .AddSingleton<IDatabaseConfig>(sp =>
                new DatabaseConfig()
                {
                    DatabaseConnectionString = dbConnectionString
                }
                )
                .BuildServiceProvider();

            IQueryExecutor executor = Schema.Create(c =>
            {
                c.RegisterQueryType(new ObjectType<TimeSeriesQuery>(d => d.Name("Query")));
                c.RegisterType<GenericObject>();
            })
            .MakeExecutable();

            var input = new TimeSeriesRequestInput()
            {
                Sensors = new List<int>() { 2, 3 },
                SpecifiedTimePeriode = false,
                TicksBackwards = 53556090190067
            };

            IReadOnlyQueryRequest request =
                QueryRequestBuilder.New()
                .SetQuery(@"query TimeSeries($input: TimeSeriesRequestInput!){
                              timeSeries(input: $input) {
                                table,
                                startDate,
                                endDate,
                                data {
                                    name,
                                    startTime
                                    interval
                                    data
                                },
                                time
                              }
                            }")
                .SetServices(serviceProvider)
                .AddVariableValue("input", input)
                .Create();

            // act
            IExecutionResult result = await executor.ExecuteAsync(request);
            //Snapshot.Match(result);
            result.MatchSnapshot();
        }


        public async Task TestSensorsQuery()
        {
            IServiceProvider serviceProvider =
                new ServiceCollection()
                .AddSingleton<IFishFarmRepository, FishFarmRepository>()
                .AddSingleton<IDatabaseConfig>(sp =>
                new DatabaseConfig()
                {
                    DatabaseConnectionString = dbConnectionString
                }
                )
                .BuildServiceProvider();

            IQueryExecutor executor = Schema.Create(c =>
            {
                c.RegisterQueryType(new ObjectType<TimeSeriesQuery>(d => d.Name("Query")));
            })
            .MakeExecutable();

            IReadOnlyQueryRequest request =
                QueryRequestBuilder.New()
                .SetQuery(@"query sensors{
                              sensors {
                                sensorTypeName,
                                sensorIds,
                                sensorColumns
                              }
                            }")
                .SetServices(serviceProvider)
                .Create();

            // act
            IExecutionResult result = await executor.ExecuteAsync(request);
            //Snapshot.Match(result);
            result.MatchSnapshot();
        }

        [Fact]
        public async Task TestTimeSeriesPeriode()
        {
            IServiceProvider serviceProvider =
                new ServiceCollection()
                .AddSingleton<IFishFarmRepository, FishFarmRepository>()
                .AddSingleton<IDatabaseConfig>(sp =>
                new DatabaseConfig()
                {
                    DatabaseConnectionString = dbConnectionString
                }
                )
                .BuildServiceProvider();

            IQueryExecutor executor = Schema.Create(c =>
            {
                c.RegisterQueryType(new ObjectType<TimeSeriesQuery>(d => d.Name("Query")));
                c.RegisterType<GenericObject>();
            })
            .MakeExecutable();

            var input = new TimeSeriesPeriodeInput()
            {
                TableName = "airsaturation_percent_",
                SensorId = 2
            };

            IReadOnlyQueryRequest request =
                QueryRequestBuilder.New()
                .SetQuery(@"query timeSeriesPeriode($input: TimeSeriesPeriodeInput){
                              timeSeriesPeriode(input: $input) {
                                from,
                                to
                              }
                            }")
                .SetServices(serviceProvider)
                .AddVariableValue("input", input)
                .Create();

            // act
            IExecutionResult result = await executor.ExecuteAsync(request);
            //Snapshot.Match(result);
            result.MatchSnapshot();
        }

    }
}
