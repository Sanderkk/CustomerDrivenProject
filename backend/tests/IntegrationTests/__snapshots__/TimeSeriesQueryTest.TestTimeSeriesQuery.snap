{
  "Data": {
    "timeSeries": {
      "table": "tension",
      "startDate": "2020-10-01T11:33:05.084+02:00",
      "endDate": "2020-10-01T11:33:05.084+02:00",
      "time": [
        "2020-08-25T23:59:59.828+02:00",
        "2020-08-25T23:59:59.578+02:00",
        "2020-08-25T23:59:59.328+02:00",
        "2020-08-25T23:59:59.078+02:00",
        "2020-08-25T23:59:58.828+02:00",
        "2020-08-25T23:59:58.578+02:00",
        "2020-08-25T23:59:58.328+02:00",
        "2020-08-25T23:59:58.078+02:00"
      ],
      "data": [],
      "numberData": [
        {
          "key": "analog_channel_0_rv009_133148",
          "value": [
            0.966184,
            0.964212,
            0.961911,
            0.966184,
            0.961911,
            0.966184,
            0.963883,
            0.966184
          ]
        },
        {
          "key": "analog_channel_1_rv004_99871",
          "value": [
            0.99182,
            0.992806,
            0.97933,
            0.987547,
            0.97933,
            0.984918,
            0.98196,
            0.997078
          ]
        },
        {
          "key": "analog_channel_2_rv011_90471",
          "value": [
            0.967827,
            0.967827,
            0.967498,
            0.967827,
            0.964869,
            0.966841,
            0.965855,
            0.966841
          ]
        },
        {
          "key": "analog_channel_3_rv010_133149",
          "value": [
            0.965526,
            0.962897,
            0.964212,
            0.966184,
            0.959282,
            0.963554,
            0.961254,
            0.961911
          ]
        },
        {
          "key": "analog_channel_4_rv005_99875",
          "value": [
            0.977358,
            0.970785,
            0.969142,
            0.96717,
            0.968156,
            0.965855,
            0.973743,
            0.971442
          ]
        },
        {
          "key": "analog_channel_5_rv008_133147",
          "value": [
            -0.000423,
            -0.001081,
            -0.000752,
            -0.000752,
            -0.000752,
            -0.000752,
            -0.001081,
            -0.000423
          ]
        },
        {
          "key": "analog_channel_6",
          "value": [
            -0.000423,
            -0.000423,
            -0.000752,
            -0.000423,
            -0.001081,
            -0.000752,
            -0.000423,
            -0.001081
          ]
        },
        {
          "key": "analog_channel_7",
          "value": [
            -0.000752,
            -0.001081,
            -0.000752,
            -0.000423,
            -0.000752,
            -0.001081,
            -0.000752,
            -0.000423
          ]
        }
      ]
    }
  },
  "Extensions": {},
  "Errors": [],
  "ContextData": {}
}
